pkg/registry/core/service/ipallocator/allocator.go

```diff
@@ -40,11 +40,18 @@ type Interface interface {
 
 var (
 	ErrFull              = errors.New("range is full")
-	ErrNotInRange        = errors.New("provided IP is not in the valid range")
 	ErrAllocated         = errors.New("provided IP is already allocated")
 	ErrMismatchedNetwork = errors.New("the provided network does not match the current range")
 )
 
+type ErrNotInRange struct {
+	ValidRange string
+}
+
+func (e *ErrNotInRange) Error() string {
+	return fmt.Sprintf("provided IP is not in the valid range. The range of valid IPs is %s", e.ValidRange)
```

Comment:
Should this print the provided IP?

Reply:
No, the provided IP is printed outside this error. See the current message:

```
Message: "Invalid value: \"127.1.2.3\": provided IP is not in the valid range",
```

The PR is expected to provide the valid IP range when we get an error of `provided IP is not in the valid range` for debugging.

Reply:
Thanks for clarifying, feel free to lgtm. We need to move forward with the reference issue quickly!